api/tokens/serializers.py

```diff
@@ -73,11 +101,54 @@ def update(self, instance, validated_data):
                 continue
             else:
                 setattr(instance, attr, value)
+        if scopes:
+            update_scopes(instance, scopes)
+
         instance.save()
         return instance
 
-def validate_requested_scopes(validated_data):
-    scopes_set = set(validated_data.get('scopes', '').split(' '))
-    for scope in scopes_set:
-        if scope not in public_scopes or not public_scopes[scope].is_public:
-            raise exceptions.ValidationError('User requested invalid scope')
+
+class ApiOAuth2PersonalTokenWritableSerializer(ApiOAuth2PersonalTokenSerializer):
+    def __init__(self, *args, **kwargs):
+        super(ApiOAuth2PersonalTokenWritableSerializer, self).__init__(*args, **kwargs)
+        request = kwargs['context']['request']
+
+        # Dynamically overriding scopes field for early versions to make scopes writable via an attribute
+        if not expect_scopes_as_relationships(request):
+            self.fields['scopes'] = ser.CharField(write_only=True, required=False)
+
+    def to_representation(self, obj, envelope='data'):
+        """
+        Overriding to_representation allows using different serializers for the request and response.
+
+        This will allow scopes to be a serializer method field if an early version, or a relationship field for a later version
+        """
+        context = self.context
+        return ApiOAuth2PersonalTokenSerializer(instance=obj, context=context).data
+
+
+def expect_scopes_as_relationships(request):
+    return StrictVersion(getattr(request, 'version', '2.0')) > StrictVersion('2.10')
+
+def update_scopes(token, scopes):
+    to_remove = token.scopes.difference(scopes)
+    to_add = scopes.difference(token.scopes.all())
+    for scope in to_remove:
+        token.scopes.remove(scope)
+    for scope in to_add:
+        token.scopes.add(scope)
+    return
+
+def validate_requested_scopes(data):
+    if not data:
+        return []
+
+    if type(data) != list:
+        data = data.split(' ')
+    scopes = ApiOAuth2Scope.objects.filter(name__in=data)
+    if len(scopes) != len(data):
+        raise exceptions.NotFound
```

Comment:
Raising a NotFound if you send in a scope that used to be in the db. Previously, a scope that wasn't listed in public scopes would throw a 400.